src/deprecated.jl

```diff
@@ -0,0 +1,111 @@
+@deprecate(
+    JLSOFile(
+        version::VersionNumber,
+        julia::VersionNumber,
+        format::Symbol,
+        compression::Symbol,
+        image::String,
+        pkgs::Dict{String, VersionNumber},
+        objects::Dict{String, Vector{UInt8}},
+    ),
+    JLSOFile(
+        version,
+        julia,
+        format,
+        compression,
+        image,
+        Pkg.TOML.parse.(_upgrade_env(pkgs))...,
+        Dict{Symbol, Vector{UInt8}}(Symbol(k) => v for (k, v) in objects),
+    )
+)
+
+@deprecate JLSOFile(data; kwargs...) JLSOFile(:data => data; kwargs...)
+@deprecate(
+    JLSOFile(data::Dict{String, <:Any}; kwargs...),
+    JLSOFile(Dict(Symbol(k) => v for (k, v) in data); kwargs...)
+)
+
+function upgrade_jlso(raw_dict::AbstractDict)
+    result = copy(raw_dict)
+    version = version_number(result["metadata"]["version"])
+
+    while version < v"3"
+        result = upgrade_jlso(result, Val(Int(version.major)))
+        version = version_number(result["metadata"]["version"])
+    end
```

Comment:
I love this bit

Reply:
You mean the iterative format upgrading?

Reply:
Should we have a debug message that dumps each upgrade state?

Reply:
> You mean the iterative format upgrading?

yes.

> Should we have a debug message that dumps each upgrade state?

Not til we need it?